server/midddleware/forwarder.go

```diff
@@ -86,9 +123,98 @@ func forwarderUnaryServerInterceptor() grpc.UnaryServerInterceptor {
 		txCtx := api.GetTransaction(ctx, req.(proto.Message))
 
 		if txCtx != nil && txCtx.GetOrigin() != types.MyOrigin {
-			return forwardRequest(ctx, txCtx, info.FullMethod, req.(proto.Message))
+			iface, err := forwardRequest(ctx, txCtx, info.FullMethod, req.(proto.Message))
+			log.Err(err).Str("method", info.FullMethod).Msg("forwarded request")
+			return iface, err
 		}
 
 		return handler(ctx, req)
 	}
 }
+
+func proxyHandler(_ interface{}, serverStream grpc.ServerStream) error {
+	fullMethodName, ok := grpc.MethodFromServerStream(serverStream)
+	if !ok {
+		return status.Errorf(codes.Internal, "lowLevelServerStream not exists in context")
+	}
+
+	outgoingCtx, backendConn, req, resp, err := proxyDirector(serverStream.Context(), fullMethodName)
+	if err != nil {
+		return err
+	}
+
+	clientCtx, clientCancel := context.WithCancel(outgoingCtx)
+	defer clientCancel()
+
+	clientStream, err := grpc.NewClientStream(clientCtx, &grpc.StreamDesc{ServerStreams: true, ClientStreams: true}, backendConn, fullMethodName)
+	if err != nil {
+		return err
+	}
+
+	s2cErrChan := forwardServerToClient(serverStream, req, clientStream)
+	c2sErrChan := forwardClientToServer(clientStream, resp, serverStream)
+
+	for i := 0; i < 2; i++ {
+		select {
+		case s2cErr := <-s2cErrChan:
+			if s2cErr == io.EOF {
+				_ = clientStream.CloseSend()
+			} else {
+				clientCancel()
```

Comment:
Don't we need to always cancel?

Reply:
actually, you already have a defer above, do you need this?

Reply:
good catch. removing.